# 1. IMPORTING TENSOFLOW DEEP LEARNING LIBRARIES

In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout
from tensorflow.keras.optimizers import Adam

# 2. TRAINING WITHOUT RESAMPLING - SMOTETOMEK

In [1]:
X = pd.read_csv("X_creditcard.csv")
y = pd.read_csv("y_creditcard.csv")

# Split into train/test
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

# Build Keras model
model = Sequential([
    Dense(64, activation='relu', input_shape=(X_train.shape[1],)),
    Dropout(0.3),
    Dense(32, activation='relu'),
    Dense(1, activation='sigmoid')  # binary output
])

model.compile(optimizer=Adam(learning_rate=0.001),
              loss='binary_crossentropy',
              metrics=['accuracy'])

# Train
history = model.fit(X_train, y_train,
                    validation_split=0.2,
                    epochs=20,
                    batch_size=32,
                    verbose=1)

# Evaluate
y_pred_probs = model.predict(X_test)
y_pred = (y_pred_probs > 0.5).astype("int32")

print("Accuracy:", accuracy_score(y_test, y_pred))
print("\nConfusion Matrix:\n", confusion_matrix(y_test, y_pred))
print("\nClassification Report:\n", classification_report(y_test, y_pred))

Epoch 1/20
5697/5697 [==============================] - 14s 2ms/step - loss: 0.0061 - accuracy: 0.9992 - val_loss: 0.0033 - val_accuracy: 0.9994
Epoch 2/20
5697/5697 [==============================] - 12s 2ms/step - loss: 0.0039 - accuracy: 0.9994 - val_loss: 0.0033 - val_accuracy: 0.9994
Epoch 3/20
5697/5697 [==============================] - 12s 2ms/step - loss: 0.0036 - accuracy: 0.9994 - val_loss: 0.0028 - val_accuracy: 0.9994
Epoch 4/20
5697/5697 [==============================] - 13s 2ms/step - loss: 0.0035 - accuracy: 0.9994 - val_loss: 0.0029 - val_accuracy: 0.9994
Epoch 5/20
5697/5697 [==============================] - 12s 2ms/step - loss: 0.0032 - accuracy: 0.9994 - val_loss: 0.0028 - val_accuracy: 0.9993
Epoch 6/20
5697/5697 [==============================] - 12s 2ms/step - loss: 0.0031 - accuracy: 0.9995 - val_loss: 0.0030 - val_accuracy: 0.9993
Epoch 7/20
5697/5697 [==============================] - 13s 2ms/step - loss: 0.0031 - accuracy: 0.9994 - val_loss: 0.0028 - val_ac

# 3. TRAINING WITHOUT RESAMPLING - SMOTETOMEK

In [2]:
X_train = pd.read_csv("X_resampled.csv")
y_train = pd.read_csv("y_resampled.csv")
X_test = pd.read_csv("X_test.csv")
y_test = pd.read_csv("y_test.csv")

# Build Keras model with Resampled Dataset

model = Sequential([
    Dense(64, activation='relu', input_shape=(X_train.shape[1],)),
    Dropout(0.3),
    Dense(32, activation='relu'),
    Dense(1, activation='sigmoid')  # binary output
])

model.compile(optimizer=Adam(learning_rate=0.001),
              loss='binary_crossentropy',
              metrics=['accuracy'])

# Train
history = model.fit(X_train, y_train,
                    validation_split=0.2,
                    epochs=20,
                    batch_size=32,
                    verbose=1)

# Evaluate
y_pred_probs = model.predict(X_test)
y_pred = (y_pred_probs > 0.5).astype("int32")

print("Accuracy:", accuracy_score(y_test, y_pred))
print("\nConfusion Matrix:\n", confusion_matrix(y_test, y_pred))
print("\nClassification Report:\n", classification_report(y_test, y_pred))

Epoch 1/20
11373/11373 [==============================] - 17s 1ms/step - loss: 0.0376 - accuracy: 0.9871 - val_loss: 0.0055 - val_accuracy: 0.9996
Epoch 2/20
11373/11373 [==============================] - 16s 1ms/step - loss: 0.0092 - accuracy: 0.9976 - val_loss: 0.0044 - val_accuracy: 0.9986
Epoch 3/20
11373/11373 [==============================] - 16s 1ms/step - loss: 0.0075 - accuracy: 0.9980 - val_loss: 0.0016 - val_accuracy: 1.0000
Epoch 4/20
11373/11373 [==============================] - 16s 1ms/step - loss: 0.0066 - accuracy: 0.9984 - val_loss: 0.0017 - val_accuracy: 1.0000
Epoch 5/20
11373/11373 [==============================] - 16s 1ms/step - loss: 0.0056 - accuracy: 0.9987 - val_loss: 0.0163 - val_accuracy: 0.9888
Epoch 6/20
11373/11373 [==============================] - 17s 1ms/step - loss: 0.0052 - accuracy: 0.9987 - val_loss: 0.0036 - val_accuracy: 0.9991
Epoch 7/20
11373/11373 [==============================] - 17s 1ms/step - loss: 0.0047 - accuracy: 0.9988 - val_loss: 0

# 4. SUMMARY

This is a quick experiment on training a fraud detection model using TensorFlow. Compared to scikit-learn, it demonstrates faster performance and improved predictive accuracy, building on the previous two notebooks.

Note that only the confusion matrix and classification report are used for evaluation, as these metrics are sufficient for assessing fraud detection performance.

## WITH UNDERSAMPLING

Before applying undersampling, the model maintains a consistent accuracy of 99.99% across 20 epochs. 
It achieves an F1 score of 0.83, indicating a reasonably balanced predictive performance despite the minority class representing only 0.17% of the dataset.

## WITHOUT UNDERSAMPLING

1. The total rows increased from 284,807 to 454,902 rows (63% increased).
2. Regardless of undersampling of minority class, the model still performs consistently at 99.99% accuracy too.
3. F1 score of minority class was from 0.83 to 0.79 is not significant difference. We can clearly infer that deep learning performs very well in classifying frauds.

# 5. CONCLUSION

With faster training using tensorflow vs Scikit-Learn and a better metric results, we can use in the deployment. It was due that our dataset is relatively large with 455k total rows which is strength of deep learning.

# 6. TO KEEP IN MIND

We have not yet applied k‑fold validation in splitting our training and testing process. Also, hyperparameter tuning is not also implemented due to current compute limitations. 
However, with access to more powerful CPU/GPU resources, our constraints will be lifted. 
In addition, we have not trained alternative models such as XGBoost or LightGBM. 
The plan is to implement this inference pipeline in an industry setting.